containerd: Image delete fixes and cleanup

dmcgowan opened the pull request:
- addresses various tests in https://github.com/moby/moby/issues/46760

Ensure that when removing an image, an image is checked consistently against the images with the same target digest.
Update logic of determining which references are the same to fix failing tests.
Add unit testing around delete.

dmcgowan committed:
Add resolve all images and unit test

Add single resolve function to get a consistent list of images matching
the same digest.

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Update cleanup logic to use resolve all images

Ensure that when removing an image, an image is checked consistently
against the images with the same target digest. Add unit testing around
delete.

Signed-off-by: Derek McGowan <derek@mcg.dev>

rumpl commented on daemon/containerd/image_delete.go:
```diff
@@ -205,19 +239,61 @@ func sortParentsByAffinity(parents []imageWithRootfs) {
 }
 
-// isSingleReference returns true if there are no other images in the
+// getSameReferences returns true if there are no other images in the
```
This method doesn't return a boolean any more, could you update the docs?

vvoland commented on daemon/containerd/image_delete.go:
```diff
@@ -205,19 +239,61 @@ func sortParentsByAffinity(parents []imageWithRootfs) {
+				}
+			} else {
+				// Ignore names which do not parse
+				log.G(ctx).WithError(err).WithField("image", ref.Name).Info("failed to parse image name, ignoring")
```
Use `WithFields`?

dmcgowan replied:
I don't usually use `WithFields` for errors and the error key is not aliased in the `log` package. There might have been a place I used `WithField` twice and should update to `WithFields` though.

vvoland commented on daemon/containerd/image_delete_test.go:
```diff
@@ -0,0 +1,219 @@
+				}
+			}
+
+			//func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, force, prune bool) ([]imagetypes.DeleteResponse, error) {
```
Do we need this comment?

vvoland commented on daemon/containerd/soft_delete.go:
```diff
@@ -17,23 +17,18 @@ const imageNameDanglingPrefix = "moby-dangling@"
 	// From this point explicitly ignore the passed context
 	// and don't allow to interrupt operation in the middle.
 
+	// If no explicit image to delete, just return and keep remaining refereneces
+	if img == nil {
+		return nil
+	}
+
```
If nil img doesn't do anything, maybe we should just accept non nilable `containerdimages.Image`?

vvoland reviewed (commented):
Just posting a couple of comments/nits.

dmcgowan commented on daemon/containerd/image_delete.go:
```diff
@@ -53,123 +53,157 @@ import (
 	defer func() {
-		if err := i.unleaseSnapshotsFromDeletedConfigs(compatcontext.WithoutCancel(ctx), possiblyDeletedConfigs); err != nil {
-			log.G(ctx).WithError(err).Warn("failed to unlease snapshots")
+		// TODO(dmcgowan): Consider leasing configs directly rather than snapshots to make this unnecessary...
```
Should I update the comment here for that?

vvoland replied:
Opened a ticket: https://github.com/moby/moby/issues/46858

I think you can remove that TODO completely now.

vvoland commented on daemon/containerd/image_test.go:
```diff
@@ -0,0 +1,280 @@
+			// Get ubuntu images with default "latest" tag
+			lookup: "ubuntu",
+			img:    &testImages[0],
+			all:    []images.Image{testImages[0], testImages[1]},
```
Instead of referencing the test image by their index, could we name them?

```go
ubuntuLatest :=  images.Image{
	Name:   "docker.io/library/ubuntu:latest",
	Target: desc(10),
}
ubuntuLatestDigested :=  images.Image{
	Name:   "docker.io/library/ubuntu:latest"@" + digestFor(10).String(),
	Target: desc(10),
}
testImages := []images.Image{
		ubuntuLatest,
		ubuntuLatestDigested
...
```

```suggestion
			all:    []images.Image{ubuntuLatest, ubuntuLatestDigested},
```

vvoland commented on daemon/containerd/image_delete.go:
```diff
@@ -53,123 +53,156 @@ import (
+	var imgID image.ID
+	if img == nil {
+		if len(all) == 0 {
+			return []imagetypes.DeleteResponse{}, nil
```
This should be unreachable, as `resolveAllReferences` would return `ErrImageDoesNotExist`? Shouldn't this also return `ErrImageDoesNotExist` to match that behavior?

https://github.com/moby/moby/blob/04526ea0fe85acdf99cf6d77bf065c72892a641e/daemon/containerd/image_delete_test.go#L30-L33

dmcgowan replied:
It definitely should be unreachable but if it is an error is easier to follow than a panic or no-op.

vvoland commented on daemon/containerd/image_delete.go:
```diff
@@ -205,19 +238,68 @@ func sortParentsByAffinity(parents []imageWithRootfs) {
-		if !isDanglingImage(ref) && ref.Name != img.Name {
-			return false, nil
+	for _, ref := range imgs {
+		if !isDanglingImage(ref) {
```
This assumes that all `imgs` have the same `Target`, right? Should we mention that in the Godoc (or check that at the beginning and return error if the assumption is broken), or it's highly unlikely that this function might be used in some other case?

vvoland commented on daemon/containerd/image.go:
```diff
@@ -341,3 +339,131 @@ func (i *ImageService) getAllImagesWithRepository(ctx context.Context, ref refer
+	var dgst digest.Digest
+	var img *containerdimages.Image
+
+	if truncatedID.MatchString(refOrID) {
+		if d, ok := parsed.(reference.Digested); ok {
+			if cimg, err := i.images.Get(ctx, d.String()); err == nil {
+				img = &cimg
+				dgst = d.Digest()
+				if cimg.Target.Digest != dgst {
+					// Ambiguous image reference, use reference name
+					log.G(ctx).WithField("image", refOrID).WithField("target", cimg.Target.Digest).Warn("digest reference points to image with a different digest")
+					dgst = cimg.Target.Digest
+				}
+			} else if !cerrdefs.IsNotFound(err) {
+				return nil, nil, convertError(err)
+			} else {
+				dgst = d.Digest()
+			}
```
(just a note when reviewing, wasn't sure which cases exactly would hit that path)

For the `^(sha256:)?([a-f0-9]{4,64})$` there are few distinct reference cases (`refOrID`):
1. `sha256:abcdef7890123456789012345678901234567890123456789012345678901234` (algo prefixed full digest)
2. `sha256:abce` (algo prefixed truncated digest)
3. `abcdef7890123456789012345678901234567890123456789012345678901234` (unprefixed full digest)
4. `abce` (unprefixed truncated digest)

This is what the reference is resolved to in each case (`reference.ParseAnyReference`):
1. `docker.io/library/sha256:abcdef7890123456789012345678901234567890123456789012345678901234`
  type: reference.taggedReference
2. `docker.io/library/sha256:abce`
  type: reference.taggedReference
3. `sha256:abcdef7890123456789012345678901234567890123456789012345678901234`
  type: reference.digestReference
4. `docker.io/library/abce`
  type: reference.repository


(https://go.dev/play/p/FU-Ysh-I2cI)

dmcgowan replied:
1 Seems a bit surprising. I can add more unit tests around these cases to lessen the surprise.

vvoland reviewed (commented):
Looks good overall. Just a few comments/questions from me. Don't think any of those is a blocker though.

vvoland commented:
Oh, also noticed this breaks the `TestRemoveByDigest` test.
The thing it checks is that you should be able to delete the image using the named and digested reference (taken from `RepoDigests`).

```bash
$ docker images
REPOSITORY   TAG       IMAGE ID       CREATED        SIZE
busybox      latest    3fbc63216742   4 months ago   6.09MB

$ docker inspect busybox -f {{.RepoDigests}}
[busybox@sha256:3fbc632167424a6d997e74f52b878d7cc478225cffac6bc977eedfe51c7f4e79]

# This should work
$ docker rmi busybox@sha256:3fbc632167424a6d997e74f52b878d7cc478225cffac6bc977eedfe51c7f4e79
Error response from daemon: No such image: busybox@sha256:3fbc632167424a6d997e74f52b878d7cc478225cffac6bc977eedfe51c7f4e79
```

dmcgowan commented:
@vvoland I'll add more unit tests around that case. I don't think the current tests match all the format of references which get created.

dmcgowan commented on daemon/containerd/image.go:
```diff
@@ -341,3 +339,131 @@ func (i *ImageService) getAllImagesWithRepository(ctx context.Context, ref refer
+			if !cerrdefs.IsNotFound(err) {
+				return nil, nil, convertError(err)
+			}
+			return nil, nil, images.ErrImageDoesNotExist{Ref: parsed}
```
Related to the remove by repo digest test, this is the case that gets hit. There is currently a unit test that checks for this condition. When a digest is provided we can probably fall back to similar logic as when just a hash is provided. The tricky part is whether to return a matched image or not, since in this case there was no actual matched image and trying to avoid any sort of fuzzy matching with whats in containerd. However I think it makes sense to support the `repo@digest` case even if it is new functionality and doesn't represent any specific image which has been created.

dmcgowan committed:
Add support for removing repo and digest

When repo and digest is provided, remove all references within a
repository for the given digest.

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan commented:
Updated with functionality to match images by repo + digest on rmi.

rumpl commented:
A lot of tests are failing now in integration-cli with `assertion failed: error is not nil: Error response from daemon: No such image: busybox:latest: [d411442a41e1e] failed to download busybox` https://github.com/moby/moby/actions/runs/7177758758?pr=46840#summary-19545421070

dmcgowan commented:
Updated, @rumpl late night commit, I missed the zero case. I added a unit test for that now as well.

Still looking at TestDockerCLIRmiSuite/TestRmiUntagHistoryLayer  but that isn't holding this PR back.

vvoland commented on daemon/containerd/image.go:
```diff
@@ -352,3 +350,132 @@ func imageFamiliarName(img containerdimages.Image) string {
+	}
+	if errors.Is(err, errInconsistentData) {
+		if retries, ok := ctx.Value(errInconsistentData).(int); !ok || retries < 3 {
+			log.G(ctx).WithField("retry", retries).WithField("ref", refOrID).Info("image changed during lookup, retrying")
```
nit
```suggestion
			log.G(ctx).WithFields(log.Fields{"retry": retries, "ref": refOrID}).Info("image changed during lookup, retrying")
```

vvoland commented on daemon/containerd/service.go:
```diff
@@ -6,15 +6,17 @@ import (
 	"github.com/docker/docker/container"
 	daemonevents "github.com/docker/docker/daemon/events"
-	"github.com/docker/docker/daemon/images"
+	daemonimages "github.com/docker/docker/daemon/images"
```
Let's keep the aliases consistent:
```suggestion
	dimages "github.com/docker/docker/daemon/images"
```

vvoland commented on daemon/containerd/image_delete_test.go:
```diff
@@ -0,0 +1,289 @@
+	"github.com/containerd/log/logtest"
+	"github.com/docker/docker/container"
+	daemonevents "github.com/docker/docker/daemon/events"
+	dockerimages "github.com/docker/docker/daemon/images"
```
Let's keep the aliases consistent:
```suggestion
	dimages "github.com/docker/docker/daemon/images"
```

vvoland commented on daemon/containerd/image_test.go:
```diff
@@ -0,0 +1,298 @@
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/log/logtest"
+	"github.com/distribution/reference"
+	dockerimages "github.com/docker/docker/daemon/images"
```
Let's keep the aliases consistent:
```suggestion
	dimages "github.com/docker/docker/daemon/images"
```

vvoland commented on daemon/containerd/image_delete.go:
```diff
@@ -53,123 +54,180 @@ import (
+		}
+		imgID = image.ID(all[0].Target.Digest)
+		var named reference.Named
+		if imageRef != string(imgID) {
```
Shouldn't this also check for truncated ID (and the dangling image)?
Like here:
https://github.com/moby/moby/blob/5c48d187eeb356f5617a3ffda5adfcfef5d5b986/daemon/containerd/image_delete.go#L107-L108

If so, we should probably extract that to a function an call it here.

dmcgowan replied:
Could check not a prefix, no need to check dangling since it would already hit this branch as not a hash or hash prefix.

dmcgowan replied:
Actually just `!isImageIDPrefix(imgID.String(), imageRef)` is enough here since that covers the equal case already

vvoland commented on daemon/containerd/image.go:
```diff
@@ -352,3 +350,132 @@ func imageFamiliarName(img containerdimages.Image) string {
+	return err
+}
+
+// resolveAllReferences resolves the reference name or ID to an image and returns all the images with
+// the same target.
+func (i *ImageService) resolveAllReferences(ctx context.Context, refOrID string) (*containerdimages.Image, []containerdimages.Image, error) {
```
I think it would be worth to highlight that the function can return nil as the first argument (`*containerdimages.Image`) while returning a non empty image slice (`[]containerdimages.Image`) and a nil error. This can only happen if the `refOrID` doesn't match to any image name, only the image target digest?

dmcgowan replied:
Added a longer section explaining the output. Gofmt seemed to really want me to go big with the formatting but think it is clearer now.

vvoland commented on daemon/containerd/image_delete.go:
```diff
@@ -53,123 +54,180 @@ import (
+					return nil, err
+				}
 			}
+			i.logImageEvent(sameRef[0], familiarRef, events.ActionUnTag)
```
Shouldn't the event be logged for every ref from `sameRef`?

dmcgowan replied:
Probably, I'll update and see if any of the tests complain. Either I was worried that the log output would differ here or wasn't quite sure that it would matter.

vvoland commented on daemon/containerd/image_delete.go:
```diff
@@ -53,123 +54,180 @@ import (
+					return nil, err
+				}
+			}
+			i.logImageEvent(*img, familiarRef, events.ActionUnTag)
```
Shouldn't the event be logged for every ref from `sameRef`?

thaJeztah commented on daemon/containerd/image.go:
```diff
@@ -352,3 +350,151 @@ func imageFamiliarName(img containerdimages.Image) string {
+		name := reference.TagNameOnly(named).String()
+
+		cimg, err := i.images.Get(ctx, name)
+		if err != nil {
+			if !cerrdefs.IsNotFound(err) {
+				return nil, nil, convertError(err)
+			}
+			if dgst == "" {
+				return nil, nil, images.ErrImageDoesNotExist{Ref: parsed}
+			}
+		} else {
```
Just a quick drive-by comment (didn't look at the whole logic); here we are _NOT_ returning for a `NotFound` error if `dgst == ""`, correct?

If that's intentional, it may be worth putting a comment in that if-branch

dmcgowan replied:
Added comment

dmcgowan committed:
Add dangling image reference on delete when last image has children

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan commented:
Updated with comment and added a commit which fixes `TestRmiUntagHistoryLayer`.

The fix for the untag is to re-add the dangling image before deleting the last tag when there are children and the rmi is not forced. I think this matches the intended logic around dangling image, another label scan right before delete is not optimal but we might be able to address that with switching to gc labels in 2.0.

vvoland commented on daemon/containerd/image_delete.go:
```diff
@@ -5,13 +5,16 @@ import (
 	"strings"
+	"time"
 
+	"github.com/containerd/containerd/errdefs"
```
Linter is unhappy:
>  daemon/containerd/image_delete.go:10:2: import "github.com/containerd/containerd/errdefs" imported without alias but must be with alias "cerrdefs" according to config (importas)
	"github.com/containerd/containerd/errdefs"
	^

```suggestion
	cerrdefs "github.com/containerd/containerd/errdefs"
```

vvoland reviewed (approved):
LGTM, thanks!

thaJeztah reviewed (approved):
LGTM

thaJeztah commented on integration-cli/docker_cli_rmi_test.go:
```diff
@@ -295,7 +295,7 @@ RUN echo 2 #layer2
 	assert.ErrorContains(c, err, "")
 	assert.Assert(c, strings.Contains(out, cID[:12]))
-	assert.Assert(c, strings.Contains(out, "(must force)"))
+	assert.Assert(c, strings.Contains(out, "(must force)") || strings.Contains(out, "(must be forced)"))
```
Oh! My search only found one, but perhaps I did it wrong.

Yeah, they were somewhat weird error messages in either case.
